### DISCUSSION TITLE
How to Add menu / Sub menu in modules ?

### DISCUSSION THREAD

**Discussion by kyotr** [oqtane/oqtane.framework#1105] [Q&A]
I created new module, but i cant find where and how to add menus in my module ? 

Many thanks

**Comment by hishamco** [comment]
Menus are generated from the pages and thier children

**Comment by sbwalker** [comment]
If you explain what you are trying to accomplish with your module from a functionality perspective it will be easier for community members to offer solutions on how to implement it using Oqtane.

**Reply by kyotr** [reply]
What i want to accomplish is that i want to declare my menus in my module in a manner that when i compile my module and i  install it, menus will appear automatically in Root menu or in any position i want.

let say that you have a module named HR.Module ; and you want users to have access to menu HR > Employees out of the box (without adding page from admin dashboard )

later, i develop another module Project.Module , and i want to add menu Task under HR ;  user will see now two menu :   HR  > Tasks  and HR > Employee

![image](https://user-images.githubusercontent.com/22255109/107825567-9ef60a00-6d83-11eb-9d36-fc6a22454e87.png)

the only way i found is to insert menus using sql scripts 
 
 i am a beginner , maybe i am missing something ?


Thanks

**Reply by hishamco** [reply]
I think you could use `PageRepository` APIs to add your pages, this way they will show up automatically in the navigation

**Reply by kyotr** [reply]
Thank you for your answer , do you have any example or documentation to follow ? it will help me alot 
Regards.

**Reply by hishamco** [reply]
The source is great reference, simply use the [PageController](https://github.com/oqtane/oqtane.framework/blob/dev/Oqtane.Server/Controllers/PageController.cs) from your module like a normal REST APIs

**Reply by sbwalker** [reply]
@kyotr  Something to consider... Oqtane is a multi-tenant framework which means that you can have many different Sites with completely different purposes that are part of the same installation. A custom module could be applicable to all Sites or it might only be applicable to a specific Site. So I am curious how you envision dealing with this during module installation - as module installation occurs at the framework level, not at the Site level.

**Reply by sbwalker** [reply]
@kyotr one possible solution you could use to accomplish your goal is the IInstallable interface. If you look at the Blog module as a reference ( https://github.com/oqtane/oqtane.blogs/blob/master/Server/Manager/BlogManager.cs ) you will see there is an Install(Tenant tenant, string version) method. Your custom module can include whatever logic you want in this method ( not only executing SQL scripts ). Since this is executed on the server you can utilize the framework Repository methods to add Pages/Modules/etc...

**Comment by sbwalker** [comment]
@kyotr I created a module which implements the IInstallable interface as an example. Basically the concept is that you can create a "migration" module whose responsibility is to configure your site by adding pages, modules, etc... The IInstallable interface is based on versioning... so you can add incremental modifications for each release by incrementing the version number. I used the Module Creator to create the module and then I removed all of the unnecessary code. What I am left with is the following:

```
/Client
  _Imports.razor
  AssemblyInfo.cs
  Index.razor - default razor component contains no logic
  ModuleInfo.cs - IModule interface needed for Version, ReleaseVersions, ServerManagerType properties
/Package
  release.cmd
  debug.cmd
/Server
  /Manager
    MigrationManager.cs - IInstallable implementation
```
The MigrationManager contains the interface implementation. You need to use the ServiceScopeFactory in order to get the services from the DI container.

```
using Oqtane.Models;
using Oqtane.Infrastructure;
using Oqtane.Repository;
using Microsoft.Extensions.DependencyInjection;
using Oqtane.Shared;
using System.Collections.Generic;
using System.Linq;
using Oqtane.Extensions;

namespace Siliqon.Migration.Manager
{
    public class MigrationManager : IInstallable
    {
        private readonly IServiceScopeFactory _serviceScopeFactory;

        public MigrationManager(IServiceScopeFactory serviceScopeFactory)
        {
            _serviceScopeFactory = serviceScopeFactory;
        }

        public bool Install(Tenant tenant, string version)
        {
            switch (version)
            {
                case "1.0.0":
                    using (var scope = _serviceScopeFactory.CreateScope())
                    {
                        var pageTemplates = new List<PageTemplate>();
                        pageTemplates.Add(new PageTemplate
                        {
                            Name = "Test",
                            Parent = "",
                            Path = "test",
                            Icon = Icons.Badge,
                            IsNavigation = true,
                            IsPersonalizable = false,
                            PagePermissions = new List<Permission>
                            {
                                new Permission(PermissionNames.View, RoleNames.Admin, true),
                                new Permission(PermissionNames.View, RoleNames.Everyone, true),
                                new Permission(PermissionNames.Edit, RoleNames.Admin, true)
                            }.EncodePermissions(),
                            PageTemplateModules = new List<PageTemplateModule>
                            {
                                new PageTemplateModule
                                {
                                    ModuleDefinitionName = "Oqtane.Modules.HtmlText, Oqtane.Client", Title = "Test", Pane = "Content",
                                    ModulePermissions = new List<Permission>
                                    {
                                        new Permission(PermissionNames.View, RoleNames.Admin, true),
                                        new Permission(PermissionNames.View, RoleNames.Everyone, true),
                                        new Permission(PermissionNames.Edit, RoleNames.Admin, true)
                                    }.EncodePermissions(),
                                    Content = "Test"
                                }
                            }
                        });

                        var siteState = scope.ServiceProvider.GetRequiredService<SiteState>();
                        siteState.Alias = new Alias { TenantId = tenant.TenantId };
                        var siteRepository = scope.ServiceProvider.GetRequiredService<ISiteRepository>();
                        List<Site> sites = siteRepository.GetSites().ToList();
                        foreach (Site site in sites)
                        {
                            siteRepository.CreatePages(site, pageTemplates);
                        }
                    }
                    break;
            }
            return true;
        }

        public bool Uninstall(Tenant tenant)
        {
            return true;
        }


    }
}
```
